### PR TITLE
New LISAv2 feature : Replaceable test parameters

### PR DESCRIPTION
**Pull request by iamshital**
Syntax :
```
RunTests.ps1 -Platform <Platform>
             -Area <Area>
             .
             .
             .
             -CustomParameters "ReplaceThis1=ReplaceWith1;ReplaceThis2=ReplaceWith2"
```
Note: Multiple replace operations can be done using sign - semicolumn (";").
Replaceable parameters needs to be defined in ".\XML\Other\ReplaceableTestParameters.xml" before using them.

### CODE REVIEW

File: Libraries/Framework.psm1 [hunk 1]
```diff
@@ -84,6 +84,39 @@ Function ValidateParameters()
...
+	if ($CustomParameters)
+	{
+		LogMsg "Checking custom parameters"
+		$CustomParameters = $CustomParameters.Trim().Trim(";").Split(";")
+		foreach ($CustomParameter in $CustomParameters)
+		{
+			$CustomParameter = $CustomParameter.Trim()
+			$ReplaceThis = $CustomParameter.Split("=")[0]
+			$ReplaceWith = $CustomParameter.Split("=")[1]
+			$OldValue = ($ReplacableTestParameters.ReplaceableTestParameters.Parameter | Where `
```
Comment (juhlee-microsoft) [inline Libraries/Framework.psm1]: Suggest you should run PSScriptAnalyzer before you merge. It should be 'Where-Object' instead of 'where'.

**Comment by iamshital** [comment]
Execution is verified.
--------------------
**Execution with custom parameters**
```
.\Runtests.ps1 -CustomParameters 'DPDK_SOURCE_URL=https://fast.dpdk.org/rel/dpdk-18.05.tar.xz;DPDK_TEST_DURATION_IN_SECONDS=60' -EnableTelemetry -TestLocation 'centralus' -RGIdentifier 'AUTOR7RAWm' -TestPlatform 'Azure' -StorageAccount 'ExistingStorage_Standard' -UpdateGlobalConfigurationFromSecretsFile -XMLSecretFile '****' -ARMImageName 'RedHat RHEL 7-RAW latest' -TestNames 'PERF-DPDK-SINGLE-CORE-PPS-DS4' -CoreCountExceededTimeout '172800' -ForceDeleteResources -UseManagedDisk
```
```
09/05/2018 21:08:26 : [INFO ] ---------------------------------
09/05/2018 21:08:26 : [INFO ] Authenticating Azure PS session..
09/05/2018 21:08:32 : [INFO ] Current Subscription : 2cd20493-xxxx-xxxx-xxxx-ab95b63d82c4.
09/05/2018 21:08:32 : [INFO ] ---------------------------------
09/05/2018 21:08:32 : [INFO ] XmlSecrets set as global variable.
09/05/2018 21:08:32 : [INFO ] Validating XML Files from C:\LISAv2\WXDV67285 folder recursively...
09/05/2018 21:08:32 : [INFO ] Collected PERF-DPDK-SINGLE-CORE-PPS-DS4
09/05/2018 21:08:32 : [INFO ] Checking custom parameters
09/05/2018 21:08:32 : [INFO ] Custom Parameter: DPDK_SOURCE_URL=https://fast.dpdk.org/rel/dpdk-18.08.tar.xz --> https://fast.dpdk.org/rel/dpdk-18.05.tar.xz
09/05/2018 21:08:32 : [INFO ] Custom Parameter: DPDK_TEST_DURATION_IN_SECONDS=120 --> 60
09/05/2018 21:08:32 : [INFO ] Custom parameter(s) are ready to be injected along with default parameters, if any.
09/05/2018 21:08:32 : [INFO ] DPDK_SOURCE_URL=https://fast.dpdk.org/rel/dpdk-18.05.tar.xz injected into C:\LISAv2\WXDV67285\TestConfiguration.xml
09/05/2018 21:08:32 : [INFO ] DPDK_TEST_DURATION_IN_SECONDS=60 injected into C:\LISAv2\WXDV67285\TestConfiguration.xml
09/05/2018 21:08:32 : [INFO ] Auto created C:\LISAv2\WXDV67285\TestConfiguration.xml validated successfully.
```

**Execution without custom parameters**
```
.\Runtests.ps1 -EnableTelemetry -TestLocation 'centralus' -RGIdentifier 'AUTOR7RAWm' -TestPlatform 'Azure' -StorageAccount 'ExistingStorage_Standard' -UpdateGlobalConfigurationFromSecretsFile -XMLSecretFile '****' -ARMImageName 'RedHat RHEL 7-RAW latest' -TestNames 'PERF-DPDK-SINGLE-CORE-PPS-DS4' -CoreCountExceededTimeout '172800' -ForceDeleteResources -UseManagedDisk
```

```
09/05/2018 21:08:50 : [INFO ] ---------------------------------
09/05/2018 21:08:50 : [INFO ] Authenticating Azure PS session..
09/05/2018 21:08:55 : [INFO ] Current Subscription : 2cd20493-xxxx-xxxx-xxxx-ab95b63d82c4.
09/05/2018 21:08:55 : [INFO ] ---------------------------------
09/05/2018 21:08:55 : [INFO ] XmlSecrets set as global variable.
09/05/2018 21:08:56 : [INFO ] Validating XML Files from C:\LISAv2\XDOF72417 folder recursively...
09/05/2018 21:08:56 : [INFO ] Collected PERF-DPDK-SINGLE-CORE-PPS-DS4
09/05/2018 21:08:56 : [INFO ] DPDK_SOURCE_URL=https://fast.dpdk.org/rel/dpdk-18.08.tar.xz injected into C:\LISAv2\XDOF72417\TestConfiguration.xml
09/05/2018 21:08:56 : [INFO ] DPDK_TEST_DURATION_IN_SECONDS=120 injected into C:\LISAv2\XDOF72417\TestConfiguration.xml
09/05/2018 21:08:56 : [INFO ] Auto created C:\LISAv2\XDOF72417\TestConfiguration.xml validated successfully.
```

File: Libraries/Framework.psm1 [hunk 2]
```diff
@@ -84,6 +84,39 @@ Function ValidateParameters()
 	}	
 }
 
+Function InjectReplaceableTestParameters($XmlConfigFilePath)
```
[the remark this answers is not in the document]
Comment (iamshital) [reply Libraries/Framework.psm1]: Sure. Will fix this.

File: XML/ReplaceableTestParameters.xml [hunk 3]
```diff
@@ -0,0 +1,46 @@
+<!--
```
[the remark this answers is not in the document]
Comment (iamshital) [reply XML/ReplaceableTestParameters.xml]: Make sense. Will do!

File: Libraries/Framework.psm1 (hunk 1 above)
Comment (iamshital) [reply Libraries/Framework.psm1]: Will do!

**Comment by chvalean** [comment]
@iamshital @juhlee-microsoft 
so any parameter that one will want to modify, must be actually first defined in XML/Other/ReplaceableTestParameters.xml?

This makes it a little bit cumbersome, because you first have to actually define it in that xml, then be able to trigger Runtests.ps1 with the changed parameter.

**Comment by iamshital** [comment]
@chvalean , This is completely optional. But yes, if you want to make some parameter "Replaceable" you need to maintain it's default values somewhere. Ju and I had discussion over this yesterday, about how can be add it to existing XML files. But, in future, the list of the "replaceable" parameters are are bound to increase. And just by looking at single file, one can understand - whether a parameter is replaceable or not.

**Comment by iamshital** [comment]
@chvalean And, I believe, controlling a feature from a centralised location, always benefits in long run.

**Pull request merged**